src/builder/context.rs

```diff
@@ -112,26 +117,102 @@ impl<'a> Context<'a> {
         };
     }
 
-    pub fn add_cache_dir(&mut self, path: &Path, name: String)
+    pub fn add_cache_dir(&mut self, path: &Path, name: &str)
         -> Result<(), String>
     {
+        let entry = match self.cache_dirs.entry(path.to_path_buf()) {
+            Entry::Vacant(entry) => entry,
+            Entry::Occupied(_) => return Ok(()),
+        };
+
+        let cache_dir = Path::new("/vagga/cache").join(&name);
+        if !cache_dir.exists() {
+            try_msg!(
+                Dir::new(&cache_dir).create(),
+                "Error creating cache dir: {err}"
+            );
+        }
+
         let path = path.strip_prefix("/")
             .map_err(|_| format!("cache_dir must be absolute: {:?}", path))?;
-        if self.cache_dirs.insert(path.to_path_buf(), name.clone()).is_none() {
-            let cache_dir = Path::new("/vagga/cache").join(&name);
-            if !cache_dir.exists() {
-                try_msg!(Dir::new(&cache_dir).create(),
-                     "Error creating cache dir: {err}");
-            }
-            let path = Path::new("/vagga/root").join(path);
-            try_msg!(Dir::new(&path).recursive(true).create(),
-                 "Error creating cache dir: {err}");
-            clean_dir(&path, false)?;
-            try_msg!(BindMount::new(&cache_dir, &path).mount(),
-                "mount cache dir: {err}");
-            self.mounted.push(path);
+        let path = Path::new("/vagga/root").join(path);
+        try_msg!(
+            Dir::new(&path).recursive(true).create(),
+            "Error creating cache dir: {err}"
+        );
+        clean_dir(&path, false)?;
+
+        try_msg!(
+            BindMount::new(&cache_dir, &path).mount(),
+            "mount cache dir: {err}"
+        );
+
+        self.mounted.push(path);
+        entry.insert(CachedDir { src: cache_dir, overlay_dir: None });
+        Ok(())
+    }
+
+    pub fn add_cache_dir_overlay(&mut self, path: &Path, name: &str)
+        -> Result<(), String>
+    {
+        let entry = match self.cache_dirs.entry(path.to_path_buf()) {
+            Entry::Vacant(entry) => entry,
+            Entry::Occupied(_) => return Ok(()),
+        };
+
+        let overlay_dir = Path::new("/vagga/container").join(name);
+        let work_dir = overlay_dir.join("work");
+        let upper_dir = overlay_dir.join("upper");
```

Comment:
I don't think `upper` is a good name here. By "upper" in the previous discussion I meant upper from `/vagga/container/root` which happens to be `/vagga/container`.

Name should be something like `tmp_cache` or something.

Also we should probably have a subfolder, so that we can add more layered things, here:
```/vagga/container/tmp_cache/apt```

Reply:
`upper` is just an overlayfs terminology. So anyone who have read overlayfs docs knows what it is for.

I like `/vagga/container/cache`.

Reply:
Yes it's upper for overlayfs, but is not clear what it is upper for. `cache` is fine.